.github/workflows/check-code-quality.yml

```
@@ -31,3 +31,32 @@ jobs:
       uses: ludeeus/action-shellcheck@master
       with:
         severity: style
+
+    - name: Check for broken links
+      run: | 
+        touch valid_urls.txt redirected_urls.txt invalid_urls.txt
+        find . -type f \( -name "*.txt" -o -name "*.md" -o -name "*.py" -o -name "*.rst" \) | 
+          xargs grep -Eo "\b(https?|ftp):\/\/[-A-Za-z0-9+&@#\/%?=~_|$!:,.;]*[A-Za-z0-9+&@#\/%=~_|$]" | 
```

I don't think forward slash needs to be escaped? According to `man grep` on my machine:
```
   Basic vs Extended Regular Expressions
       In  basic  regular expressions the meta‐characters ?, +, {, |, (, and )
       lose their special meaning; instead use the  backslashed  versions  \?,
       \+, \{, \|, \(, and \).
```

```suggestion
          xargs grep -Eo "\b(https?|ftp)://[-A-Za-z0-9+&@#/%?=~_|$!:,.;]*[A-Za-z0-9+&@#/%=~_|$]" | 
```

---

Separately: do we actually have any ftp links, and do we want to check them? Might be overkill.